akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -0,0 +1,372 @@
+                  } else {
+                    if (context.log.isTraceEnabled) {
+                      context.log.traceN(
+                        "Got reply for old tx id [{}] (current [{}]) for pid [{}], ignoring",
+                        transactionId,
+                        stateForPid.currentTransactionId,
+                        pid)
+                    }
+                  }
+              }
+              Behaviors.same
+
+            case JournalProtocol.WriteMessageFailure(message, error, transactionId) =>
+              val pid = message.persistenceId
+              val sequenceNr = message.sequenceNr
+              perPidWriteState.get(pid) match {
+                case None =>
+                  context.log.debugN(
+                    "Got write error reply for event with no previous state for pid, ignoring (pid [{}], seq nr [{}], tx id [{}])",
+                    pid,
+                    sequenceNr,
+                    transactionId)
+                case Some(state) =>
+                  // write failure could be re-delivery, we need to check
+                  state.waitingForReply.get(sequenceNr) match {
+                    case None =>
+                      context.log.debugN(
+                        "Got write error reply for event with no waiting request for seq nr, ignoring (pid [{}], seq nr [{}], tx id [{}])",
+                        pid,
+                        sequenceNr,
+                        transactionId)
+                    case Some(_) =>
+                      // quite likely a re-delivery of already persisted events, the whole batch will fail
+                      // check highest seqnr and see if we can ack events
+                      if (!state.writeErrorHandlingInProgress && state.currentTransactionId == transactionId) {
+                        perPidWriteState =
+                          perPidWriteState.updated(pid, state.copy(writeErrorHandlingInProgress = true))
+                        context.ask(
+                          journal,
+                          (replyTo: ActorRef[JournalProtocol.Response]) =>
+                            JournalProtocol.ReplayMessages(0L, 0L, 1L, pid, replyTo.toClassic)) {
+                          case Success(JournalProtocol.RecoverySuccess(highestSequenceNr)) =>
+                            MaxSeqNrForPid(pid, highestSequenceNr, error.getMessage)
+                          case Success(unexpected) =>
+                            throw new IllegalArgumentException(
+                              s"Got unexpected reply from journal ${unexpected.getClass} for pid $pid")
+                          case Failure(exception) =>
+                            throw new RuntimeException(
+                              s"Error finding highest sequence number in journal for pid $pid",
+                              exception)
+                        }
+                      } else {
+                        if (context.log.isTraceEnabled)
```

Comment:
if not needed, shouldn't happen too often, and it's only saving an array allocation